waterbox/opera/bizhawk.cpp

```diff
@@ -0,0 +1,402 @@
+
+	//// Getting memory areas
+	/** 2 = wram, 3 = vram*/
+
+	// WRAM
+	_memoryAreas.wram = (uint8_t*)retro_get_memory_data(RETRO_MEMORY_SYSTEM_RAM);
+	_memorySizes.wram = retro_get_memory_size(RETRO_MEMORY_SYSTEM_RAM);
+
+	// VRAM
+	_memoryAreas.vram = (uint8_t*)retro_get_memory_data(RETRO_MEMORY_VIDEO_RAM);
+	_memorySizes.vram = retro_get_memory_size(RETRO_MEMORY_VIDEO_RAM);
+
+	// Getting av info
+	struct retro_system_av_info info;
+	retro_get_system_av_info(&info);
+	printf("3DO Framerate: %f\n", info.timing.fps);
+
+	return true;
+}
+
+ECL_EXPORT void opera_get_video(int& w, int& h, int& pitch, uint8_t*& buffer)
+{
+	buffer = (uint8_t*)_videoBuffer;
+	w = _videoWidth;
+	h = _videoHeight;
+	pitch = _videoPitch;
+}
+
+ECL_EXPORT void FrameAdvance(MyFrameInfo* f)
+{
+	 // Setting inputs
+		_port1Value = f->port1;
+		_port2Value = f->port2;
+
+  //printf("Mouse X%d(%d), Y%d(%d), L%d, M%d, B%d\n", _port1Value.mouse.posX, _port1Value.mouse.dX, _port1Value.mouse.posY, _port1Value.mouse.dY, _port1Value.mouse.leftButton, _port1Value.mouse.middleButton, _port1Value.mouse.rightButton);
+		//fflush(stdout);
+
+  // Checking for changes in NVRAM
+		_nvramChanged = false;
+
+  // Checking for lag frames
+		_isLagFrame = true;
+
+		// If resetting, do it now. Otherwise, running a single frame
+		if (f->isReset == 1) retro_reset();
+		else retro_run();
+
+		// Setting if lag frame
+		f->base.Lagged = _isLagFrame ? 1 : 0;
+
+		// Setting video buffer
+		f->base.Width = _videoWidth;
+		f->base.Height = _videoHeight;
+		memcpy(f->base.VideoBuffer, _videoBuffer, sizeof(uint32_t) * _videoWidth * _videoHeight);
+
+		// Setting audio buffer
+		f->base.Samples = _audioSamples;
+		memcpy(f->base.SoundBuffer, _audioBuffer, _audioSamples * sizeof(int16_t) * _CHANNEL_COUNT);
+}
+
+ECL_EXPORT void GetMemoryAreas(MemoryArea *m)
+{
+	int memAreaIdx = 0;
+
+	m[memAreaIdx].Data  = _memoryAreas.wram;
+	m[memAreaIdx].Name  = "Work RAM";
+	m[memAreaIdx].Size  = _memorySizes.wram;
+	m[memAreaIdx].Flags = MEMORYAREA_FLAGS_WORDSIZE1 | MEMORYAREA_FLAGS_WRITABLE | MEMORYAREA_FLAGS_PRIMARY;
+	memAreaIdx++;
+
+	m[memAreaIdx].Data  = _memoryAreas.vram;
+	m[memAreaIdx].Name  = "Video RAM";
+	m[memAreaIdx].Size  = _memorySizes.vram;
+	m[memAreaIdx].Flags = MEMORYAREA_FLAGS_WORDSIZE1 | MEMORYAREA_FLAGS_WRITABLE;
+	memAreaIdx++;
+}
+
+void (*InputCallback)();
+ECL_EXPORT void SetInputCallback(void (*callback)())
+{
+	InputCallback = callback;
+}
+
+bool _sram_changed = false;
+ECL_EXPORT bool sram_changed() { return _nvramChanged; }
+ECL_EXPORT int get_sram_size() { return NVRAM_SIZE; }
+ECL_EXPORT uint8_t* get_sram_buffer() { return (uint8_t*) NVRAM; }
+ECL_EXPORT void get_sram(uint8_t* sramBuffer)
+{
+	 if (NVRAM == NULL) return;
+		memcpy(sramBuffer, get_sram_buffer(), get_sram_size());
+}
+
+ECL_EXPORT void set_sram(uint8_t* sramBuffer)
+{
+	 if (NVRAM == NULL) opera_nvram_init(NVRAM,NVRAM_SIZE);
+		memcpy(get_sram_buffer(), sramBuffer, get_sram_size());
+}

```

Comment:
For some reason `retro_get_memory_data` does not seem to expose this NVRAM (via `RETRO_MEMORY_SAVE_RAM` for example) but perhaps that would be the better place to expose this to instead of writing an own function for it.

You could then also expose this NVRAM as memory domain with the `MEMORYAREA_FLAGS_SAVERAMMABLE` flag and I believe the frontend will then automatically handle saveram for you.

Reply:
fixed